lib/src/local_working_copy.rs

```diff
@@ -1008,6 +1011,14 @@ impl TreeState {
                                 }
                             }
                         }
+                    } else if start_tracking_matcher.visit(&path).is_nothing()
+                        && current_tree.path_value(&path)?.is_absent()
```

Comment:
nit: maybe this can be `file_states.is_empty()`? It's a bit odd that we test `current_tree` here whereas scan `file_states` for git-ignored directory.

Maybe we can also combine these two to optimize directory walk.

Reply:
> nit: maybe this can be `file_states.is_empty()`? It's a bit odd that we test `current_tree` here whereas scan `file_states` for git-ignored directory.

Good idea. Done.

> Maybe we can also combine these two to optimize directory walk.

The gitignore matcher is a different kind of matcher, so I don't see how. Or do you just mean that we could skip asking the gitignore matcher if `file_states.is_empty()`? That makes sense.

Reply:
> The gitignore matcher is a different kind of matcher, so I don't see how. Or do you just mean that we could skip asking the gitignore matcher if `file_states.is_empty()`? That makes sense.

I mean these two `if`s can be combined.
```rust
                    if git_ignore.matches(&path.to_internal_dir_string())
                        || start_tracking_matcher.visit(&path).is_nothing()
                    {
                        // If the whole directory is ignored, visit only paths we're already
                        // tracking.
                        for (tracked_path, current_file_state) in file_states {
```

btw, in order to warn untracked paths, we'll need to visit subdirectory even if `file_states.is_empty()`. An empty untracked directory or directories containing only ignored files shouldn't be warned.

Reply:
> btw, in order to warn untracked paths, we'll need to visit subdirectory even if `file_states.is_empty()`. An empty untracked directory or directories containing only ignored files shouldn't be warned.

Good point. I've merged the conditions and updated the comment.